### PR TITLE
Don't fail if the user is offline

### PR DESCRIPTION
**Pull request by nagilson**
Resolves https://github.com/dotnet/vscode-dotnet-runtime/issues/1866
https://github.com/dotnet/vscode-dotnet-runtime/issues/1858 (does not add api, we decide to do this automatically)

The `VersionResolver` class makes a web-request before downloading .NET to find the newest version of .NET to keep users up to date. The problem is, this resolution of which version of .NET to download occurs before when we check the pre-existing installations to see if we need to do an installation at all, because we need to resolve a version string such as '8.0' to '8.0.2'. 

If the user had downloaded the .NET before hand then went on an airplane, which many have, it would not check if that install exists because it'd first see if there is a newer version to download then fail. 

This adds a check where if we are offline we check the preexisting installs first to prevent this failure from occurring.

# Previous Behavior:

It would wait for the timeout silently and then fail.
See users reporting this in https://github.com/microsoft/vscode-dotnettools/issues/1028 which this resolves. 

# New Behavior:

image

For the first 2 requests, I request '7.0' which was installed previously. It uses the old existing install and remarks to update when possible. 

For the next request, I request '8.0' which was not installed previously. It has a new warning message instead of hanging and will wait..

When connecting back to the internet, it works again.

### CODE REVIEW

File: vscode-dotnet-runtime-extension/src/extension.ts [hunk 1]
```diff
@@ -176,13 +179,28 @@ export function activate(vsCodeContext: vscode.ExtensionContext, extensionContex
                 return existingPath;
             }
 
+            if(!(await WebRequestWorker.isOnline(timeoutValue ?? defaultTimeoutValue, globalEventStream)))
+            {
+                workerContext.acquisitionContext.architecture = workerContext.acquisitionContext.architecture ?? DotnetCoreAcquisitionWorker.defaultArchitecture();
```
Comment (smitpatel) [inline vscode-dotnet-runtime-extension/src/extension.ts:184]: Can use `??=`

File: vscode-dotnet-runtime-extension/src/extension.ts [hunk 2]
```diff
@@ -231,20 +249,29 @@ export function activate(vsCodeContext: vscode.ExtensionContext, extensionContex
...
             }
 
-            const globalInstallerResolver = new GlobalInstallerResolver(sdkContext, commandContext.version);
+            if(!(await WebRequestWorker.isOnline(timeoutValue ?? defaultTimeoutValue, globalEventStream)))
+            {
+                const existingPathOffline = await worker.getSimilarExistingInstall(workerContext);
+                if(existingPathOffline?.dotnetPath)
+                {
+                    return Promise.resolve(existingPathOffline);
+                }
```
Comment (smitpatel) [inline vscode-dotnet-runtime-extension/src/extension.ts:258]: This does not log message like the other?

File: vscode-dotnet-runtime-library/src/Acquisition/VersionUtilities.ts [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+        const event = new DotnetFeatureBandDoesNotExistError(new EventCancellationError('DotnetFeatureBandDoesNotExistError',
+            `${invalidFeatureBandErrorString}${fullySpecifiedVersion}.`),
+            getInstallFromContext(context));
+        eventStream.post(event);
+        throw event.error;
+    }
+    return patch
+}
+
+
```
Comment (smitpatel) [inline vscode-dotnet-runtime-library/src/Acquisition/VersionUtilities.ts:92]: extra line break

**Review by smitpatel — APPROVED** [review]
Please split the function to handle command in different files. It is making harder to review changes in extension.ts file as it is hard to find which function is getting modified.

File: vscode-dotnet-runtime-extension/src/extension.ts (hunk 1 above)
Comment (nagilson) [reply vscode-dotnet-runtime-extension/src/extension.ts:184]: omg thank you I forgot that operator existed 🤯

File: vscode-dotnet-runtime-extension/src/extension.ts (hunk 2 above)
Comment (nagilson) [reply vscode-dotnet-runtime-extension/src/extension.ts:258]: yikes it also doesnt set the architecture lemme refactor it out into a function

Comment (nagilson) [reply vscode-dotnet-runtime-extension/src/extension.ts:258]: good catch thank you!!

**Comment by nagilson** [comment]
Thanks for your review, you caught some important issues. For the refactor, I've made a separate issue to track this so it can be done in another PR: https://github.com/dotnet/vscode-dotnet-runtime/issues/1917

**Pull request merged**